tpu-client-next/src/connection_worker.rs

```diff
@@ -0,0 +1,220 @@
+        command_receiver: mpsc::Receiver<TransactionBatch>,
+    ) -> Self {
+        Self {
+            endpoint,
+            peer,
+            transactions_receiver: command_receiver,
+            connection: ConnectionState::NotSetup,
+            send_txs_stats: SendTransactionStats::default(),
+        }
+    }
+
+    /// Starts the main loop of the [`ConnectionWorker`]. This method manages the
+    /// connection to the peer and handles state transitions. It runs
+    /// indefinitely until the connection is closed or an unrecoverable error
+    /// occurs.
+    pub async fn run(&mut self) {
+        loop {
+            match &self.connection {
+                ConnectionState::Closing => {
+                    break;
+                }
+                ConnectionState::NotSetup => {
+                    self.create_connection(0).await;
+                }
+                ConnectionState::Active(connection) => {
+                    let Some(transactions) = self.transactions_receiver.recv().await else {
+                        info!("Transactions sender has been dropped.");
+                        self.connection = ConnectionState::Closing;
+                        continue;
+                    };
+                    self.send_transactions(connection.clone(), transactions)
+                        .await;
+                }
+                ConnectionState::Retry(num_reconnects) => {
+                    if *num_reconnects > MAX_RECONNECT_ATTEMPTS {
+                        error!(
+                            "Didn't manage to establish connection: reach max reconnect attempts."
+                        );
+                        self.connection = ConnectionState::Closing;
+                        continue;
+                    }
+                    sleep(RETRY_SLEEP_INTERVAL).await;
+                    self.reconnect(*num_reconnects).await;
+                }
+            }
+        }
+    }
+
+    /// Retrieves the statistics for transactions sent by this worker.
+    pub fn get_transaction_stats(&self) -> &SendTransactionStats {
+        &self.send_txs_stats
+    }
+
+    /// Sends a batch of transactions using the provided `connection`. Each
+    /// transaction in the batch is sent over the QUIC streams one at the time,
+    /// which prevents traffic fragmentation and shows better TPS in comparison
+    /// with multistream send. If the batch is determined to be outdated, it
+    /// will be dropped without being sent. In case of error, it doesn't retry
+    /// to send the same transactions again.
+    async fn send_transactions(&mut self, connection: Connection, transactions: TransactionBatch) {
+        let now = timestamp();
+        if now.saturating_sub(transactions.get_timestamp()) > MAX_PROCESSING_AGE_MS {
+            info!("Drop outdated transaction batch.");
+            return;
+        }
+        let mut measure_send = Measure::start("send transaction batch");
+        for data in transactions.into_iter() {
```

Comment:
Track how many transactions have already been sent and being treated, for staked connections there is a way to calculate how many parallel streams a connection can have. Try to parallize only that many transactions at once.

Reply:
There is no need to parallelize anything. Parallelizing makes things slower by creating fragmentation.

Reply:
We count how many txs have been sent in `SendTransactionStats::successfully_sent` counter. Regarding send in parallel using number of streams, it is no longer needed because on the server side will not allow us to send more than it can accept from a client with provided stake.